controllers/terminal_controller.go

```diff
@@ -1241,6 +1242,19 @@ func (r *TerminalReconciler) createOrUpdateTerminalPod(ctx context.Context, cs *
 						Effect:   corev1.TaintEffectNoSchedule,
 					})
 			}
+			if !tolerationExists(criticalAddonsKey) {
+				pod.Spec.Tolerations = append(pod.Spec.Tolerations,
+					corev1.Toleration{
+						Key:      criticalAddonsKey,
+						Operator: corev1.TolerationOpExists,
+					})
+			}
+			pod.Spec.Tolerations = append(pod.Spec.Tolerations,
+				corev1.Toleration{
+					Operator: corev1.TolerationOpExists,
+					Effect:   corev1.TaintEffectNoExecute,
+				},
+			)
```

Comment:
couldn't it be that this toleration was already added from a previous reconcile? And on each reconcile it will be appended again and again (haven't tested it though to verify)

Reply:
Yes, you are correct. I've added a way check if a toleration exists by comparing the whole tolerations struct as this specific one does not have a key. 
If the check is only done by effect and operator then this toleration will not get added if there is already a toleration with `effect: NoExecute` and `operator: Exists` but with a specific key.